consensus/hotstuff/timeoutcollector/timeout_cache.go

```diff
@@ -0,0 +1,108 @@
+package timeoutcollector
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+var (
+	// ErrRepeatedTimeout is emitted, when we receive an identical timeout object for the same block
+	// from the same voter multiple times. This error does _not_ indicate
+	// equivocation.
+	ErrRepeatedTimeout            = errors.New("duplicated timeout")
+	ErrTimeoutForIncompatibleView = errors.New("timeout for incompatible view")
+)
+
+// TimeoutObjectsCache maintains a _concurrency safe_ cache of timeouts for one particular
+// view. The cache memorizes the order in which the timeouts were received. Timeouts
+// are de-duplicated based on the following rules:
+//  * For each voter (i.e. SignerID), we store the _first_ timeout t0.
+//  * For any subsequent timeout t, we check whether t.ID() == t0.ID().
+//    If this is the case, we consider the timeout a duplicate and drop it.
+//    If t and t0 have different checksums, the voter is equivocating, and
+//    we return a model.DoubleTimeoutError.
+type TimeoutObjectsCache struct {
+	lock     sync.RWMutex
+	view     uint64
+	timeouts map[flow.Identifier]*model.TimeoutObject // signerID -> first timeout
+}
+
+// NewTimeoutObjectsCache instantiates a TimeoutObjectsCache for the given view
+func NewTimeoutObjectsCache(view uint64) *TimeoutObjectsCache {
+	return &TimeoutObjectsCache{
+		view:     view,
+		timeouts: make(map[flow.Identifier]*model.TimeoutObject),
+	}
+}
+
+func (vc *TimeoutObjectsCache) View() uint64 { return vc.view }
+
+// AddTimeoutObject stores a timeout in the cache. The following errors are expected during
+// normal operations:
+//  * nil: if the vote was successfully added
+//  * model.DoubleTimeoutError is returned if the voter is equivocating
+//  * RepeatedTimeoutErr is returned when adding a timeout for the same view from
+//    the same voter multiple times.
+//  * TimeoutForIncompatibleViewError is returned if the timeout is for a different view.
+// When AddTimeoutObject returns an error, the timeout is _not_ stored.
+func (vc *TimeoutObjectsCache) AddTimeoutObject(timeout *model.TimeoutObject) error {
+	if timeout.View != vc.view {
+		return ErrTimeoutForIncompatibleView
+	}
+	vc.lock.Lock()
+	defer vc.lock.Unlock()
+
+	// De-duplicated timeouts based on the following rules:
+	//  * For each voter (i.e. SignerID), we store the _first_  t0.
+	//  * For any subsequent timeout t, we check whether t.ID() == t0.ID().
+	//    If this is the case, we consider the timeout a duplicate and drop it.
+	//    If t and t0 have different checksums, the voter is equivocating, and
+	//    we return a model.DoubleTimeoutError.
+	firstTimeout, exists := vc.timeouts[timeout.SignerID]
+	if exists {
+		if firstTimeout.ID() != timeout.ID() {
```

Comment:
Determining equality via ID is a pretty expensive operation, as it requires full marshalling (using reflection) and a hash computation. I would expect that our ID function is _at best_ as performant as Go's `reflect.DeepEqual`, which is roughly 100 slower than a custom implementation of an equality function based on [this benchmark](https://www.sobyte.net/post/2021-12/struct-compare/).

Once we have signer indices, it will be straight forward to implement equal functions for QCs, TCs, and thereby TimeoutObjects. I think this would reduce the surface area of a spamming attack here. 

```suggestion
		// TODO: once we have signer indices, implement Equals methods for QC, TC 
		// and TimeoutObjects, to avoid the comparatively very expensive ID computation. 
		if firstTimeout.ID() != timeout.ID() {
```